docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,215 @@
+When a transaction is selected and reaped, it is removed from this index and
+from the `<sender:*Tx>` mapping.
+
+We define `Tx` as the following data structure:
+
+```go
+type Tx struct {
+  // Tx represents the raw binary transaction data.
+  Tx []byte
+
+  // Priority defines the transaction's priority as specified by the application
+  // in the ResponseCheckTx response.
+  Priority int64
+
+  // Sender defines the transaction's sender as specified by the application in
+  // the ResponseCheckTx response.
+  Sender string
+
+  // Index defines the current index in the priority queue index. Note, if
+  // multiple Tx indexes are needed, this field will be removed and each Tx
+  // index will have its own wrapped Tx type.
+  Index int
+}
+```
+
+### Eviction
+
+Upon successfully executing `CheckTx` for a new `Tx` and the mempool is currently
+full, we must check if there exists a `Tx` of lower priority that can be evicted
+to make room for the new `Tx` with higher priority and with sufficient size
+capacity left.
+
+If such a `Tx` exists, we find it by obtaining a read lock and sorting the
+priority queue index. Once sorted, we find the first `Tx` with lower priority and
+size such that the new `Tx` would fit within the mempool's size limit. We then
+remove this `Tx` from the priority queue index as well as the `<sender:*Tx>`
+mapping.
+
+This will require additional `O(n)` space and `O(n*log(n))` runtime complexity.
+
+### Gossiping
+
+When gossiping new transactions to peers, we first obtain a read lock on the
+priority queue index. If we have not gossiped the top priority `Tx` for the
+given peer, we return this `Tx` as the next transaction to gossip. Otherwise,
+we sort the priority queue index and traverse in descending priority order until
+we find the first non-sent `Tx` to the peer.
+
+In order to facilitate this, we must also keep an ephemeral cache, mapping from
+`sender` to a list of unique `Tx` identifiers (e.g. the `Tx` hash). A `Tx`
+identifier is removed from this cache once it is either committed in a block or
+gossiped to a peer. Note, if the `Tx` is committed in a block, this requires that
+we iterates over the entire cache, i.e. each peer's list.
+
+## Consequences
+
+### Positive
+
+- Transactions are allowed to be prioritized by the application.
+- Transactions are allowed to be gossiped by priority.
+- Allows future replace-by-priority functionality.
+
+### Negative
+
+- Additional bytes sent over the wire due to new fields added to `ResponseCheckTx`.
+- A transaction that passed `CheckTx` and entered the mempool can later be evicted
+  at a future point in time if a higher priority transaction entered while the
+  mempool was full.
+- Possible decreased throughput performance due to the necessity of various read
+  and write locks, e.g. sorting when either gossiping or evicting transactions.
+- It is possible that certain transactions broadcasted in a particular order may
```

Comment:
Right. This is the case I was highlighting in my other comments which I realized shouldn't be possible in the SDK. But if it is possible in an app, then an attacker can potentially seriously abuse it to fill the blockchain with invalid txs. 

Not sure if there's a good solution to this other than punting it to PrepareProposal from ABCI++ to figure it out

Reply:
Yeah I honestly don't think this is a major concern. These types of casually ordered txs are not common and in the cases they do happen, the entities involved just need to ensure the 1st tx has a higher fee than the 2nd.

In any case, ABCI++ will help us here so I'm not too concerned about it.

Reply:
I was more so highlighting an attack scenario. If the app does allow these dependent txs (the SDK does not, but maybe others do, but I guess they shoudln't!) then an attacker could spam the mempool with them in a way that will result in blocks with invalid txs

Reply:
I don't think its so clear that ABCI++ will help us here. Immediate execution does not seem like a given for most chains, for the same reasons we designed to not do this for the SDK. (Despite us not taking advantage of those reasons though lol - one big one is being able to mutate the underlying data store in place)

Sorting out the ordering of txs definitely feels like a mempool responsibility, which in the future may mean that 'threads' of txs that concurrently pass CheckTx have to be queued up for validator nodes.